lib/astarte_core/triggers/policy/policy.ex

```diff
@@ -0,0 +1,180 @@
+# limitations under the License.
+#
+
+defmodule Astarte.Core.Triggers.Policy do
+  use Astarte.Core.Triggers.PolicyProtobuf
+  use Ecto.Schema
+  import Ecto.Changeset
+
+  alias Astarte.Core.Triggers.Policy
+  alias Astarte.Core.Triggers.Policy.Handler
+  alias Astarte.Core.Triggers.Policy.TaggedPolicy
+
+  @required_fields [
+    :name,
+    :error_handlers,
+    :maximum_capacity
+  ]
+
+  @permitted_fields [
+    :name,
+    :maximum_capacity,
+    :retry_times,
+    :event_ttl
+  ]
+
+  @primary_key false
+  embedded_schema do
+    field :name
+    field :maximum_capacity, :integer
+    field :retry_times, :integer
+    field :event_ttl, :integer
+    embeds_many :error_handlers, Policy.Handler
+  end
+
+  def changeset(%Policy{} = policy, params \\ %{}) do
+    # TODO check minimum and maxiumim range of attributes + conditions on name
+    policy
+    |> cast(params, @permitted_fields)
+    |> validate_required(@required_fields)
+    |> validate_length(:name, max: 128)
+    |> validate_format(:name, policy_name_regex())
+    |> validate_exclusion(:name, ["astarte", "control", "device"])
+    |> validate_inclusion(:retry_times, 1..100)
+    |> validate_inclusion(:event_ttl, 1..100)
+    |> validate_inclusion(:maximum_capacity, 1..1_000_000)
+    |> cast_embed(:error_handlers, with: &Handler.changeset/2, required: true)
+    |> validate_length(:error_handlers, min: 1, max: 200)
+    |> validate_all_handlers_on_different_errors()
+    |> validate_all_discard_no_retry_times()
+    |> validate_retry_times_if_retry_strategy()
+  end
+
+  def policy_name_regex do
+    ~r/^([a-zA-Z][a-zA-Z0-9]*\.([a-zA-Z0-9][a-zA-Z0-9-]*\.)*)?[a-zA-Z][a-zA-Z0-9]*$/
+  end
+
+  defp validate_all_handlers_on_different_errors(changeset) do
+    handlers = get_field(changeset, :error_handlers, [])
+
+    if all_different_errors(handlers) do
+      changeset
+    else
+      add_error(changeset, :error_handlers, "must all handle distinct errors.")
+    end
+  end
+
+  # TODO Maybe use a computationally better comparison?
+  defp all_different_errors(handlers) do
+    errs = List.flatten(Enum.map(handlers, &Handler.error_list/1))
+    Kernel.length(errs) == Kernel.length(Enum.uniq(errs))
+  end
+
+  defp validate_all_discard_no_retry_times(changeset) do
+    handlers = get_field(changeset, :error_handlers, [])
+    retry_times = get_field(changeset, :retry_times)
+
+    if Enum.all?(handlers, &Handler.discards?/1) and retry_times != nil do
+      add_error(changeset, :retry_times, "must not be set if all errors are discarded.")
+    else
+      changeset
+    end
+  end
+
+  defp validate_retry_times_if_retry_strategy(changeset) do
+    handlers = get_field(changeset, :error_handlers, [])
+    retry_times = get_field(changeset, :retry_times)
+
+    if Enum.any?(handlers, fn h -> not Handler.discards?(h) end) and retry_times == nil do
+      add_error(changeset, :retry_times, "must be set if some events are to be retried.")
+    else
+      changeset
+    end
+  end
```

Comment:
These could be combined in a single check since `Enum.all?(enum, some_fun)` == `Enum.any?(enum, not some_fun)`

```elixir
all_discards = Enum.all?(handlers, &Handler.discards?/1)

cond do
  all_discards and retry_times != nil ->
    ...
  not all_discards and retry_time == nil ->
    ...
  true ->
    changeset
end
```